File: Diameter-of-Binary-Tree.go
func maxmax(a,b int) int{
    if(a>b){
        return a;
    }
    return b
}

func getTreeHeight(root *TreeNode) int {
    if(root == nil){
        return 0
    }
    return maxmax(getTreeHeight(root.Left),getTreeHeight(root.Right)) + 1
}

func diameterOfBinaryTree(root *TreeNode) int {
    if(nil == root){return 0;}
        
    leftHeight := getTreeHeight(root.Left);
    rightHeight := getTreeHeight(root.Right);
        
    leftDiameter := diameterOfBinaryTree(root.Left);
    rightDiameter := diameterOfBinaryTree(root.Right);
        
    return maxmax(leftHeight+rightHeight, maxmax(leftDiameter,rightDiameter));
}

File: Diameter-of-Binary-Tree.cpp
// 递归函数临时变量达到4个就会报栈溢出

class Solution {
public:
    int result;
    int getTreeHeight(TreeNode* root) {
        if(!root)return 0;
        int left = getTreeHeight(root->left);
        int right = getTreeHeight(root->right);
        
        if(left+right > result){
            result = left+right;
        }
        
        return max(left,right)+1;
        
    }
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root){return 0;}
        result =0;
        getTreeHeight(root);
        return result;

    }
};
